src/dev.rs

```diff
@@ -0,0 +1,293 @@
+//! Tools for developing circuits.
+
+use ff::Field;
+
+use crate::{
+    arithmetic::{FieldExt, Group},
+    plonk::{Any, Assignment, Circuit, Column, ConstraintSystem, Error},
+    poly::{EvaluationDomain, LagrangeCoeff, Polynomial},
+};
+
+#[derive(Debug)]
+struct Cell(usize, usize);
+
+/// The reasons why a particular circuit is not satisfied.
+#[derive(Debug, PartialEq)]
+pub enum VerifyFailure {
+    /// A gate was not satisfied for a particular row.
+    Gate { gate_index: usize, row: usize },
+    /// A lookup input did not exist in its corresponding table.
+    Lookup { lookup_index: usize, row: usize },
+    /// A permutation did not preserve the original value of a cell.
+    Permutation {
+        perm_index: usize,
+        column: usize,
+        row: usize,
+    },
+}
+
+/// Assembly for permutation argument
+pub struct PermutationAssembly {
+    pub mapping: Vec<Vec<Vec<(usize, usize)>>>,
+    aux: Vec<Vec<Vec<(usize, usize)>>>,
+    sizes: Vec<Vec<Vec<usize>>>,
+}
+
+/// A test
+pub struct MockProver<F: Group> {
+    n: u32,
+    domain: EvaluationDomain<F>,
+    cs: ConstraintSystem<F>,
+
+    // The fixed cells in the circuit, arranged as [column][row].
+    fixed: Vec<Polynomial<F, LagrangeCoeff>>,
+    // The advice cells in the circuit, arranged as [column][row].
+    advice: Vec<Polynomial<F, LagrangeCoeff>>,
+    // The aux cells in the circuit, arranged as [column][row].
+    aux: Vec<Polynomial<F, LagrangeCoeff>>,
+
+    permutations: PermutationAssembly,
+}
+
+impl<F: Field + Group> Assignment<F> for MockProver<F> {
+    fn assign_advice(
+        &mut self,
+        column: crate::plonk::Column<crate::plonk::Advice>,
+        row: usize,
+        to: impl FnOnce() -> Result<F, crate::plonk::Error>,
+    ) -> Result<(), crate::plonk::Error> {
+        *self
+            .advice
+            .get_mut(column.index())
+            .and_then(|v| v.get_mut(row))
+            .ok_or(Error::BoundsFailure)? = to()?;
+
+        Ok(())
+    }
+
+    fn assign_fixed(
+        &mut self,
+        column: crate::plonk::Column<crate::plonk::Fixed>,
+        row: usize,
+        to: impl FnOnce() -> Result<F, crate::plonk::Error>,
+    ) -> Result<(), crate::plonk::Error> {
+        *self
+            .fixed
+            .get_mut(column.index())
+            .and_then(|v| v.get_mut(row))
+            .ok_or(Error::BoundsFailure)? = to()?;
+
+        Ok(())
+    }
+
+    fn copy(
+        &mut self,
+        permutation: usize,
+        left_column: usize,
+        left_row: usize,
+        right_column: usize,
+        right_row: usize,
+    ) -> Result<(), crate::plonk::Error> {
```

Comment:
This seems to be duplicated from the normal prover; does it need to be?

Reply:
I have de-duplicated this.